sway-lsp/src/server.rs

```diff
@@ -217,3 +223,310 @@ impl LanguageServer for Backend {
         ))
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use serde_json::json;
+    use std::{env, fs::File, io::Write};
+    use tower::{Service, ServiceExt};
+
+    use super::*;
+    use futures::stream::StreamExt;
+    use tower_lsp::jsonrpc::{self, Request, Response};
+    use tower_lsp::LspService;
+
+    // Simple sway script used for testing LSP capabilites
+    const SWAY_PROGRAM: &str = r#"script;
+
+use std::*;
+
+/// A simple Particle struct
+struct Particle {
+    position: [u64; 3],
+    velocity: [u64; 3],
+    acceleration: [u64; 3],
+    mass: u64,
+}
+
+impl Particle {
+    /// Creates a new Particle with the given position, velocity, acceleration, and mass
+    fn new(position: [u64; 3], velocity: [u64; 3], acceleration: [u64; 3], mass: u64) -> Particle {
+        Particle {
+            position: position,
+            velocity: velocity,
+            acceleration: acceleration,
+            mass: mass,
+        }
+    }
+}
+
+fn main() {
+    let position = [0, 0, 0];
+    let velocity = [0, 1, 0];
+    let acceleration = [1, 1, 0];
+    let mass = 10;
+    let p = ~Particle::new(position, velocity, acceleration, mass);
+}
+"#;
```

Comment:
Fun to see what could be a `nannou` test in  `sway` code :smile:

Reply:
Come to think of it, would it be possible to use the actual `sway` repo examples for testing? I'm imagining  you could do something like:

```rust
mod example {
    pub const HELLO_WORLD: &str = include_str!("../path/to/hello_world/main.rs");
    pub const FIZZBUZZ: &str = ...;
    // ...
}

pub const SWAY_PROGRAMS: &[&str] = &[
    example::HELLO_WORLD,
    example::FIZZBUZZ,
    // ...
];
```

and then where relevant, test each of them in a loop? We may be able to catch more issues that way?

Reply:
Nice, I like this idea. The current test file is handy for directly comparing the results with rust analyzer, but we can implement a feature like this for more generalized future tests.